library/core/src/fmt/mod.rs

```diff
@@ -2281,16 +2282,51 @@ impl<T: ?Sized> Pointer for &mut T {
 
 // Implementation of Display/Debug for various core types
 
+/// A local trait for pointer Debug impl so that we can have
+/// min_specialization-compliant specializations for two types of fat ptrs.
+trait PtrMetadataFmt {
+    fn fmt_ptr(&self, ptr: *const (), f: &mut Formatter<'_>) -> Result;
+}
+
+// Regular pointer / default impl
+impl<T> PtrMetadataFmt for T {
```

Comment:
You should be able to improve this in one of two ways:
* split the `()` case from the default:
  * `()` is "guaranteed simple" (but maybe specializing on the original pointee being `Sized` would be stronger)
  * the default would only be hit by custom DSTs (and unreachable today? could print some indication of metadata existing but not being understood, or just fit an `unreachable!()` in there)
* add a new **sealed** trait, and a bound for it on `Pointee::Metadata` (just like it requires `Copy`, `Eq`, etc. today)
  *  ("sealed" as in a `pub trait` in a private `mod`, not nameable outside `core` - for now, anyway)
  * the new trait could just be this PR's `PtrMetadataFmt` for now, even if we may want to change it in the future (see below)
  * no specialization should be needed, since `Pointee` would be claiming it's always implemented for valid metadata types
  * whenever we get custom DSTs, it would be a nice bonus to force the user to implement such a trait, effectively making them decide how raw pointers to their custom DST should be formatted
    * this is when we'd remove the "sealing"
    * maybe it should just be a `fmt::Debug` bound and not let the user see the data pointer or choose how *it* is printed?
    * but this part is really all a bikshed for the future

Reply:
I know exactly what you mean, in fact, I've been meaning to propose in #81513 that `Pointee::Metadata` should maybe be its own trait (and yeah initially sealed by all means), which could enable attaching arbitrary things to it that might be needed in the future, be it debug printing or something else.

I initially did try the `fmt::Debug` bound approach but I think there was issues with it, altough I don't remember off the top of my head at the moment...

Reply:
> I initially did try the `fmt::Debug` bound approach but I think there was issues with it, altough I don't remember off the top of my head at the moment...

Right, because `Debug` on `Pointee::Metadata` gets me the ability to debug-print the metadata, but in this case I need to format the whole pointer+metadata combo somehow, ideally preserving the simple "just hex number" format for simple pointers...